sycl/source/detail/program_manager/program_manager.cpp

```diff
@@ -481,10 +481,40 @@ RT::PiProgram ProgramManager::getBuiltPIProgram(
   if (Prg)
     Prg->stableSerializeSpecConstRegistry(SpecConsts);
 
-  auto BuildF = [this, &M, &KSId, &ContextImpl, &DeviceImpl, Prg, &CompileOpts,
+  // Check if root device architecture is homogeneous and we can optimize builds
+  // for sub-devices
+  DeviceImplPtr RootDevImpl = DeviceImpl;
+  while (!RootDevImpl->isRootDevice()) {
+    auto ParentDev = detail::getSyclObjImpl(
+        RootDevImpl->get_info<info::device::parent_device>());
+    if (!ContextImpl->hasDevice(ParentDev))
+      break;
+    RootDevImpl = ParentDev;
+  }
+
+  pi_bool IsRootDeviceArchHomogeneous = PI_FALSE;
+  ContextImpl->getPlugin().call<PiApiKind::piDeviceGetInfo>(
+      RootDevImpl->getHandleRef(), PI_DEVICE_INFO_HOMOGENEOUS_ARCH,
+      sizeof(pi_bool), &IsRootDeviceArchHomogeneous, nullptr);
+
+  // FIXME: the logic is modified to work around unintuitive Intel OpenCL CPU
+  // implementation behavior. Kernels created with the program built for root
+  // device can be re-used on sub-devices, but other combinations doesn't work
+  // (e.g. clGetKernelWorkGroupInfo returns CL_INVALID_KERNEL if kernel was
+  // created from the program built for sub-device and re-used either on root or
+  // other sub-device).
+  // To work around this case we optimize only one case: root device shares the
+  // same context with its sub-device(s). We built for the root device and
```

Comment:
Where is "the same context" checked? What if context just has no root-device in it, only all of its sub-devices?

Reply:
See https://github.com/intel/llvm/pull/5240/files#diff-78dd7f7ba0b6120dece1ae4ab5a09c9936ff654a1de2c31ff2dbb1fc58d90393R490. Put additional comment to emphasize.

Reply:
> What if context just has no root-device in it, only all of its sub-devices?

The optimization won't be enabled in such case.

Reply:
I see, thanks. Maybe as a future optimization we could implicitly add the root-device to the context, if >1 of it's sub-devices are there already (such that we can save on 1+ module builds). If you agree, please consider adding a TODO comment.

Reply:
I'm not sure SYCL spec allows implicitly add devices to the context implicitly created by the runtime, but I think it's **not** allowed if the context is provided by the user.

The latest patch removes following comment, which I was considering as direction for future optimizations: https://github.com/intel/llvm/pull/5240/commits/e6ca4f905caeb1cac075827234b27cdd09f18a5c#diff-78dd7f7ba0b6120dece1ae4ab5a09c9936ff654a1de2c31ff2dbb1fc58d90393L509-L511
I think it would be great if Level Zero allows us to re-use the program built for any (sub-)device and not only a root device. I tested it on Intel GPU and it works already, but again it's not guaranteed by the spec wording. In this case we don't need implicitly add the root-device to optimize the build for sub-devices.

@bashbaug, does it make sense to pursue this direction? If so, I can recover the comment.

Reply:
I can think of some cases in theory at least where a program built for one sub-device wouldn't be valid for a sibling sub-device, so this is not a safe assumption in all cases.  If we decide this is a direction we want to pursue we'd need to find a way to detect or request this behavior.